satellite/overlay/keylock.go

```diff
@@ -0,0 +1,64 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information
+
+package overlay
+
+import (
+	"sync"
+
+	"storj.io/storj/pkg/storj"
+)
+
+// KeyLock provides per-key RW locking. Locking is key-specific, meaning lock
+// contention only exists for locking calls using the same key parameter. As
+// with all locks, do not call Unlock() or RUnlock() before a corresponding
+// Lock() or RLock() call.
+//
+// Note on memory usage: internally KeyLock lazily and atomically creates a
+// separate sync.RWMutex for each key. To maintain synchronization guarantees,
+// these interal mutexes are not freed until the entire KeyLock instance is
+// freed.
+type KeyLock struct {
```

Comment:
I wonder whether there's a nicer name for this. Something like `NodeLocker` or something :D?